Accel calibration

superjax opened the pull request:
Supersedes #22

There are some things i would like to fix, but I need your help @dpkoch.

I think we should save the calibration parameters from the IMU to a file.  They are currently saved as a private member of the Imu class, and they are solved for on [line 94 of imu.cpp](https://github.com/BYU-MAGICC/fcu_io2/blob/accel_calibration/src/mavrosflight/sensors/imu.cpp#L93).  I think it should be pretty easy to save off the values in `x_` to the parameter handler, but I'm not sure how to get that working.  These same parameters should be loaded on the constructor for the `Imu` class.  That is in [lines 20-22 of imu.cpp](https://github.com/BYU-MAGICC/fcu_io2/blob/accel_calibration/src/mavrosflight/sensors/imu.cpp#L20-22).  Do you think that's possible?

After those things are going, then I think we should make calibrating the IMU a service call to `fcu_io`.  That way, it is only called when someone wants it called.

 I also got the barometer into it's own class, it has a `correct` function and the whole thing.

superjax committed:
got barometer working on new refactored system

superjax committed:
rough start at accel calibration

superjax committed:
accel is being calibrated on the ROS side

superjax committed:
got accelerometer calibration working

superjax committed:
finished moving barometer to new style

superjax commented on src/fcu_io.cpp:
```diff
@@ -61,33 +67,43 @@ void fcuIO::handle_mavlink_message(const mavlink_message_t &msg)
 void fcuIO::on_new_param_received(std::string name, double value)
-{
-  ROS_INFO("Got parameter %s with value %g", name.c_str(), value);
+{https://github.com/simondlevy/
```
I don't know where this keeps coming from

superjax committed:
ran a style update

superjax commented on src/mavrosflight/sensors/imu.cpp:
```diff
@@ -1,28 +1,116 @@
+
+      // Perform Least-Squares on the data
+      x_[i] = Amat.jacobiSvd(Eigen::ComputeFullU | Eigen::ComputeFullV).solve(Bmat);
+      ROS_INFO_STREAM("x" << i <<" = " << x_[i]); // <-- These should get saved in the parameter server
```
At this point, I would like to save the parameters in `x_` to the parameter handler

superjax commented on src/mavrosflight/sensors/imu.cpp:
```diff
@@ -1,28 +1,116 @@
+{
+  x_[0] << 0, 0;
+  x_[1] << 0, 0;
+  x_[2] << 0, 0;
```
Here, I would like to instead load the parameters from previous calibrations.

superjax replied:
Actually, I think, since this is already going to be on the naze, `fcu_io` doesn't need to know what these are unless it's doing a calibration.

superjax committed:
added lost baro files

superjax commented:
You should know, as well, that the Naze will calibrate it's own gyro biases.  That is all handled internally by the estimator.

dpkoch commented on src/fcu_io.cpp:
```diff
@@ -82,6 +86,16 @@ void fcuIO::handle_mavlink_message(const mavlink_message_t &msg)
+  case MAVLINK_MSG_ID_PARAM_VALUE: // <-- This is handled by the param_manager
+    break;
+  default:
+    ROS_ERROR("fcu_io:Got unsupported message ID %d", msg.msgid);
```
I'm just wondering if you're trying to accomplish something specific here? In my mind receiving an unhandled message isn't an error, it just means we're not interested in it. I had removed it at one point, but at the very least we should demote it to a warning.

superjax replied:
Oh, I was just trying to track down why fcu_io kept freezing.

dpkoch commented on src/fcu_io.cpp:
```diff
@@ -127,6 +141,12 @@ void fcuIO::handle_small_imu_msg(const mavlink_message_t &msg)
   temp_msg.header.stamp = imu_msg.header.stamp;
 
+  // This is so we can eventually make calibrating the IMU an external service
+  if (!imu_.calibrated)
```
Yep, an external service call would be great. We won't want to calibrate every time we start up, but I assume this is just a placeholder for now so that's great

dpkoch commented:
Okay, I think we're on the same page here. The calibration is computed on the ROS side, but stored and applied on the Naze side. You had just said you were putting it in `Imu::correct`, which was why I was concerned. This is what I had in mind.

Yeah, the current architecture is going to make it a little awkward saving the calibration to the parameters. It requires a lot of sibling-to-sibling communication among class members, which I don't love but haven't figured out how to fix yet. The cleanest solution as it is right now might be to return the parameters from the calibrate function, then in fcu_io call `mavrosflight_->param.set_param_value(meh)` to save them. Then in the `on_new_param_received` callback function listen for those parameters and pass them to the `imu_` object (we'd have to add a function to `sensors::Imu` to enable that). Not pretty, but I'm not sure I have a better idea without some refactoring. Which will need to happen someday. What do you think?

dpkoch commented:
Reading your question again, I guess you were hoping to save the parameters to a file. I was thinking we just save them to the Naze onboard parameters, that way the Naze has access to them too. At some point I want to support saving and loading the whole set of parameters to/from a file (see #21), which would include the calibration parameters, but I'm afraid that's not going to happen tonight.

dpkoch commented:
Also, just FYI, a useful Github practice is to prepend the name of pull requests like this with `[WIP]` (work in progress). It lets people know it's not ready to merge in yet (so don't merge a pull request if you see that), but still lets you have the conversation and a place to work on it.

dpkoch commented:
So that's great the estimator estimates gyro biases. However, I think we should still calibrate, if for no other reason than to stream better measurements to the computer. I imagine it couldn't hurt the estimator either though.

The calibration could be super simple though. In sensors.c we just average the first 100 or so gyro measurements at startup, then store those as biases and subtract them off from all subsequent measurements. What do you think?

superjax commented:
I wasn't feeling all the particular about the place that they were saved. I mostly just meant into the 'param handler'  which I understood to mean whatever system we use to save and load parameters.  Saving them onto the naze sounds like a great idea.

I like the idea of `[WIP]`.  I'll do that next time.  Thanks.

superjax commented:
I think that'll work.  The estimator could tweak them as it goes along, because the gyro biases do seem to wander a little bit.

dpkoch commented on src/mavrosflight/sensors/imu.cpp:
```diff
@@ -1,28 +1,116 @@
-  *yacc = msg.yacc * ACCEL_SCALE;
-  *zacc = msg.zacc * ACCEL_SCALE;
+  double temperature = ((double)msg.temperature/340.0 + 36.53);
+  *xacc = (msg.xacc - (temperature)*x_[0](0) - x_[0](1)) * ACCEL_SCALE;
```
So I'm thinking this shouldn't happen here (the temperature compensation). It needs to happen on the Naze anyway, so it seems like we should just do it there and then stream the compensated value over mavlink. That would also mean  that we don't have to load in the compensation parameters, since the Naze is taking care of it. All we have to do on the ROS side is compute a calibration and send it the Naze, then let the Naze take care of the rest. Does that make sense to you?

superjax replied:
Yeah.  I like that.  I guess the reason I was doing it here was because the data was already there, so it was easy.  I like the idea of having the naze do it all.

dpkoch commented:
Did you want me to implement the saving things to the parameter server stuff? I'm happy to do it, but it will be a couple of hours before I can work on it

superjax committed:
hooked up accel calibration to parameter server, but crashing fcu_io

superjax commented:
I already got the parameter server stuff running.  Sorta...  Sending so much to the naze all at once is causing fcu_io to crash.  I think it is the same problem we're seeing in #15.  I have started diving into the mavros stuff to try and figure it out, but I'm a little fuzzy on how the `WriteBuffer` works.  From what I understand, I don't think we should be having a problem.

superjax commented on src/fcu_io.cpp:
```diff
@@ -127,6 +142,28 @@ void fcuIO::handle_small_imu_msg(const mavlink_message_t &msg)
+        mavrosflight_->param.set_param_value("ACC_X_BIAS", 1000.0*imu_.xb());
+        mavrosflight_->param.set_param_value("ACC_Y_BIAS", 1000.0*imu_.yb());
+        mavrosflight_->param.set_param_value("ACC_Z_BIAS", 1000.0*imu_.zb());
+        mavrosflight_->param.write_params();
```
I'm pretty sure this is sorta what both you and I were thinking about how to pass the biases back over to the naze.  Do you know of any reason why this would cause mavros to crash?

dpkoch commented:
Yeah that's weird, I don't know why it's crashing. It's supposed to just buffer them. `WriteBuffer` keeps track of the output buffer for a single message, and then the `write_queue_` buffers those. If it gets a bunch of messages all at once, theoretically it will just push them all onto the queue and send them as it can. So from what I understand we shouldn't be having a problem either. I'll see if I can get some eyes on it.

superjax commented:
I'm not getting an error message, that would be fantastic if we were.  The node just freezes.

dpkoch commented:
The asynchronous reading and writing happen via a `boost::ioservice` that runs a separate thread; see Line 41 in mavlink_serial.cpp.

superjax committed:
might have found a bug

superjax commented on src/mavrosflight/mavlink_serial.cpp:
```diff
@@ -135,7 +139,7 @@ void MavlinkSerial::send_message(const mavlink_message_t &msg)
+  // is there supposed to be a condition here?
   {
     mutex_lock lock(mutex_);
     write_queue_.push_back(buffer);
```
As I said in the comment.  I'm wondering if this is freezing the code.  (by the way, the code definitely freezes here)

superjax replied:
Ah, it's a `scoped_mutex`.  I get why the brackets are there now.  So, the code is freezing at line 144 `mutex_lock lock(mutex_);`

superjax committed:
changed write_queue to be a lockfree queue, so the broken mutex business isn't a problem anymore

superjax commented:
So, I'm still not sure what was happening.  in the `async_write_end` function, it calls the `do_async_write` function, but I'm not sure that the `scoped_lock` gets destroyed. Anyway, it seemed to be freezing there, because at that point, both the `send_message` and the `do_async_write` want to lock the mutex, but it's being held by `async_write_end`.  Anyway, I bypassed the whole issue (and made the `async_write_end` function a little easier by changing the `write_buffer_` to a `boost::lockfree::queue`

superjax commented:
Alright, I think I'm ready to merge this in.  Go through and check my work, but I think that the threading stuff I modified should fix our issues, as well as #15.

As for gyro biases, I think it'll be better to either calculate the biases using the naze's estimator, or in `fcu_io` but not both.  Personally, I'm more inclined to use the naze's estimator, simply because I've seen it work really well.  The idea then is that after a 10th of a second or so, the gryo measurements coming over mavlink would already be adjusted.  If they were both calculating biases, then fcu_io might not keep up with the naze, and we would just be introducing artificial bias.

dpkoch commented:
Interesting. I'm too tired to parse this right now, but I'll look through it first thing Monday morning. Thanks for working on it! (Sorry for being slow and not contributing anything this weekend. But it's because I've been busy scuba diving, so I'm only kind of sorry :smile:)

So with the gyro biases, I agree that we shouldn't calculate them in `fcu_io`. I am a little hesitant, however, to say that we should stream gyro values corrected by the estimator's bias estimate. My reasoning is that the onboard bias estimates might make the attitude estimator perform really well (yay!), but there's no guarantee that they're the actual true bias. So if you're wanting to do estimation on the ROS side, then if you use those bias estimates it's possible that you could get some subtle weird effects that would be tough to track down. Do you think that's legitimate?

What if we just computed a gyro offset on startup (on the Naze), then used that constant value to correct the gyro readings feeding into the estimator and streaming over mavlink? Then the attitude estimator could still adjust the biases from there internally as it pleases, but it wouldn't affect what gets streamed over mavlink. If someone wants more sophisticated bias estimation on the ROS side, then they're welcome to implement it (like in the RMEKF), but we wouldn't do anything in fcu_io. What do you think about that idea?

superjax commented:
No problem about being busy.   I totally get it.  Thanks for the help you offered.  There's no rush, except that I want it working well for you guys down at REEF (gotta put our best foot forward right?)   I would like it to work well for you guys before you leave so that they think that ROSflight is a good idea.

Yeah, I'll have to think about gyro biases a little more, but I'll probably get to that on Tuesday.

dpkoch commented:
So my current best guess is that the locking issue is caused by the recursive calls to `do_async_write`. I looked at what mavros is doing these days, and they've switched to a `boost::recursive_mutex` instead of a simple `boost::mutex` like it was before. From the [boost documentation](http://www.boost.org/doc/libs/1_31_0/libs/thread/doc/recursive_mutex.html#introduction):
> These types should be used to synchronize access to shared resources when recursive locking by a single thread is likely to occur. A good example for this is when a class supplies "internal synchronization" to ensure thread-safety and a function of the class may have to call other functions of the class which also attempt to lock the mutex.

Sounds like what we have going on. I backed up a couple of commits and switched out the mutex type. I've been sitting here with fcu_io running for the last several minutes, streaming IMU at 500Hz and sending command messages at 1000Hz with no apparent issues. Before, sending commands at 500Hz would break fcu_io fairly regularly. Now data is still streaming and param functionality still works (IMU streaming did start slowing down when I had it at 1000Hz, but there's a lot going on so I'm not too worried about it).

So, if it doesn't offend you too much, I'm going to revert your last commit or two (I'll save them off on a separate branch) and implement these changes.

superjax commented:
Sounds good.  That agrees with the issues I was seeing earlier.  That is a whole lot of data to push on that serial line.  It's pretty cool that it works that well.

superjax commented:
So as for the gyros...  what if we left the gyro bias streaming stuff as a parameter?  I agree that passing the converged biases from the complementary filter may introduce artificial bias if it goes to RMEKF, but I'm wondering if it might be nice to get the zero-bias measurements for less-sophisticated stuff, (like the camera driver I'm working on).  It would be a pretty advanced parameter adjustment, but if someone were to read the documentation they would be able to decide whether or not they wanted the bias subtracted off or not.

dpkoch committed:
Reverted mavlink_serial.h and mavlink_serial.cpp to commit 4a56a4b

dpkoch committed:
Changed mavlink_serial lock guard to recursive_mutex

dpkoch commented:
Sure, as long as it's turned off by default I think that'd be fine. I'd still say to exercise caution if you decide to use them (I don't really know what your application is, is this for the rovio stuff?), but I think having the option would be good.

Question: I noticed that you changed the minimum required temperature change for calibration to be really small. Is that something you actually want in there, or is it a debugging thing (so you don't have to start with a cold IMU) that should be reverted? Perhaps we should leave it bigger to get a more accurate calibration, but add a timeout so that the calibration routine still exits if there's not a big enough temperature swing?

dpkoch committed:
Added missing calibrate callback return value

dpkoch committed:
Demoted unhandled mavlink message ID severity to 'DEBUG'

superjax commented:
Okay, I'll build that into ROSflight real quick.  I'm using it for ROVIO.  I read through the Mahoney Paper, and the Lyuapunov analysis indicates that it does actually converge to the true biases.  You know how these non-linear analyses are, 'in infinite time, they will converge' so, you're right, they may not actually be perfect ever, but I think that they are going to be really good.  In my tests with it so far, they are have been really pretty accurate.

Yeah, I was working on it at home, and I had forgotten my freon.  Let's change it back to 1 deg C, and let's put a timeout of 10 seconds, with a warning that perhaps the temperature compensation isn't perfect.

superjax commented:
Also, do you want to put in the gyro bias calculation stuff real quick?

dpkoch commented:
Were we thinking the gyro bias stuff should go in ROSflight?

For that matter, do we want the ability to do a quick accel bias calculation in ROSflight (without doing the temperature compensation; just leaving that as it is)? So it might work like this:
- For the majority of the time, the user calls the `calibrate_imu` service (doing some renaming here), that instructs ROSflight to perform an onboard bias computation for accel and gyro
- If the user is interested, they call the `calibrate_imu_temperature` service that calculates the temperature compensation values. We mostly care about the slope value here since the offset value may get changed by the above routine, but we'll take them too since we get them for free

My thought is that then we've reduced some dependency on our specific GCS tool, plus the average user doesn't need to understand that they have to start with a cold IMU and let it warm up, etc., to get a good bias calibration. They just get the default temperature compensation that should be good enough for most people anyway (like how the DMP library does it).

Thoughts?

superjax commented:
I hadn't thought of that, but I think that's a good idea.  I'll work on the ROSflight side of it

dpkoch committed:
Changed temperature compensation timeout and temperature requirement

dpkoch commented:
Cool. Sounds good, thanks. I'll get the fcu_io side working.

Some info on how we'll implement this with mavlink:
- From fcu_io I'll send a COMMAND_INT message with a command value of MAV_CMD_PREFLIGHT_CALIBRATION. I'll set the flags to do a gyro calibration and an accel calibration.
- On the ROSflight side, you'll need to add a case to the switch statement in `mavlink_handle_msg_command_int()` in src/mavlink_receive.c. Call the appropriate calibration routines (they should be called separately to match the mavlink interface that lets you do that), then set the result flag within your case.

Does that make sense? Also, do you think it would be better to have separate `calibrate_accel` and `calibrate_gyro` services, or have them both in one `calibrate_imu` service?

superjax commented:
I was thinking that they should probably both happen at the same time.  Is there any reason why we would have them separate?

dpkoch commented:
No they're probably fine together; it was just an option. I suppose conceivably you could calibrate gyros if the vehicle is sitting still but not level, but I'm not sure having them separate would be all that useful. 

I'd say implement it in ROSflight to support having them separate (you just have to check the fields of the command message and call the calibration routines separately. Go to https://pixhawk.ethz.ch/mavlink/ and search for MAV_CMD_PREFLIGHT_CALIBRATION for a description of the fields). But for now I'll just implement the one service call in fcu_io that does them both. Good?

superjax commented:
So, weird thing.  The MAVlink documentation has this:
![image](https://cloud.githubusercontent.com/assets/7344014/16593439/819ca2a4-42a2-11e6-9049-59e13d0971a6.png)
But the code shows this
```c
typedef struct __mavlink_command_int_t
{
 float param1; /*< PARAM1, see MAV_CMD enum*/
 float param2; /*< PARAM2, see MAV_CMD enum*/
 float param3; /*< PARAM3, see MAV_CMD enum*/
 float param4; /*< PARAM4, see MAV_CMD enum*/
 int32_t x; /*< PARAM5 / local: x position in meters * 1e4, global: latitude in degrees * 10^7*/
 int32_t y; /*< PARAM6 / local: y position in meters * 1e4, global: longitude in degrees * 10^7*/
 float z; /*< PARAM7 / z position: global: altitude in meters (relative or absolute, depending on frame.*/
 uint16_t command; /*< The scheduled action for the mission item. see MAV_CMD in common.xml MAVLink specs*/
 uint8_t target_system; /*< System ID*/
 uint8_t target_component; /*< Component ID*/
 uint8_t frame; /*< The coordinate system of the COMMAND. see MAV_FRAME in mavlink_types.h*/
 uint8_t current; /*< false:0, true:1*/
 uint8_t autocontinue; /*< autocontinue to next wp*/
} mavlink_command_int_t;
```

I assume you'll be using `param1` and `x` from the struct?

dpkoch commented:
Huh, that's annoying. Yeah I suppose I will. The documentation for the message itself (https://pixhawk.ethz.ch/mavlink/#COMMAND_INT) shows that x doubles for param5, etc. So that's kind of ugly/stupid, but I guess it's what we're stuck with.

dpkoch committed:
Made IMU calibration more object-oriented

dpkoch committed:
Fixed type casting bug in param manager

Fixed bug where setting a param value would not cast the requested value to the param type before recording it as the new value.

dpkoch committed:
Renamed IMU temperature calibration service

dpkoch committed:
Added service and handling for IMU bias calibration command

dpkoch commented:
Okay well... I think we're almost there. But not quite.

I've been testing with a rev2 and a rev5 flip32, and have been seeing a couple of issues:

1. Starting the onboard calibration zeros the gyros and x and y accel really well, but the z accel always has an offset (about 1 m/s^2).
2. For the temperature calibration
 1. 10 seconds is not long enough to get sufficient temperature variation even right after plugging it in, so we'll have to figure something else out for the timeout
 2. The scaling is way off. The computed biases translate to about 65 m/s^2

This should probably also be linked to BYU-MAGICC/ROSflight2#29. The fact that the scaling is different between board revisions is not helping either.

I'm taking a quick lunch break now, but we'll need to figure this out before the calibration will be of much use. It may involve some redesign on how we deal with the scaling. I'm not sure what the best path forward is; I'm open to suggestions if you have them.

dpkoch commented:
See BYU-MAGICC/ROSflight2#38 for one possible path forward...

superjax committed:
fixed the temperature calibration routine to actually work

superjax commented:
So I just got this working. 

Like an idiot, I was mixing up what went into A and B in the Ax = B equation.  We were getting the equation for temperature as a function of accel, not the other way around.   :-P...  Also, I set it up to reset the previous calibration before trying again.  We were getting a compouding of calibrations that was just making things break.  

Here is my rqt_plot after calibrating
![image](https://cloud.githubusercontent.com/assets/7344014/16631026/18f97ad2-437a-11e6-8a55-3f00f7f2dd10.png)

dpkoch committed:
Modified completion criteria, cleaned out debug code

superjax commented:
It's because the calibration is performed on the naze.  Fcu_io is getting pre-calibrated values when it collects the data, but when it pushes the new modified values, those old values are lost, so the offsets will be wrong.

dpkoch commented:
Ah, right. Makes sense.

dpkoch commented:
Okay, I think we're about ready to merge this in? I think the last thing we need to do is decide whether we're moving to floating point as per BYU-MAGICC/ROSflight2#38. If we want to, we could either merge the accel_calibration_float branch into this one, then both into master, or do them one at a time. I'm fine with either way, let's just wrap this guy up!

dpkoch commented:
This pull request closes #15.

superjax commented:
I think we are ready to merge it in, but I am not sure that I am totally convinced about the floating point stuff.  It seems to me that if we can figure out why Breezy is not configuring your MPU6050 right, then the problem is solved, and we don't have to worry about the bandwidth loss streaming IMU.

dpkoch commented:
Alright let's merge it in as is and leave the floating point discussion for later. I think that there are other reasons why it'd be a good idea, one being greatly improved readability and another being that fcu_io would no longer have to know about the internal settings of the Naze to do its job. But let's continue that discussion in BYU-MAGICC/ROSflight2#38 and merge this in here.
